packages/client-engine-runtime/src/interpreter/renderQuery.ts

```diff
@@ -1,11 +1,23 @@
 import { ArgType, SqlQuery } from '@prisma/driver-adapter-utils'
 
-import { isPrismaValuePlaceholder, PrismaValue, QueryPlanDbQuery } from '../QueryPlan'
-import { renderQueryTemplate } from './renderQueryTemplate'
+import { Fragment, isPrismaValuePlaceholder, PlaceholderFormat, PrismaValue, QueryPlanDbQuery } from '../QueryPlan'
 import { ScopeBindings } from './scope'
 
-export function renderQuery({ query, params }: QueryPlanDbQuery, scope: ScopeBindings): SqlQuery {
-  const substitutedParams = params.map((param) => {
+export function renderQuery(dbQuery: QueryPlanDbQuery, scope: ScopeBindings): SqlQuery {
+  switch (dbQuery.type) {
+    case 'rawSql':
+      return renderRawSql(dbQuery.sql, substituteParams(dbQuery.params, scope))
+
+    case 'templateSql':
+      return renderTemplateSql(dbQuery.fragments, dbQuery.placeholder, substituteParams(dbQuery.params, scope))
+
+    default:
+      return dbQuery // never
```

Comment:
```suggestion
      assertNever(dbQuery.type, 'Invalid query type')
```
so that we get a compile error if we don't handle any newly added case (and an additional failsafe at runtime)

Reply:
`asserNever` is defined in `utils.ts` in this package

Reply:
Eh, I had the runtime fail-safe there, then removed it as "unnecessary". Looks like it is actually good practice then.

Reply:
I'm mostly interested in compile time file-safe, e.g. I'd be okay with just
```ts
  default:
    dbQuery.type satisfies never
```

`assertNever` combines both this and a runtime exception (that should be unreachable at run time) — this is indeed redundant but doesn't hurt.

Reply:
I guess your variant also works because otherwise there would be a type error due to `dbQuery` not being assignable to the return type of the function, but it's a little bit more brittle than explicitly requiring that `dbQuery.type` is `never` in this branch (e.g. what if the types changed in the future and the type of `dbQuery` became a subtype of the return type?)